src/commcare_cloud/ansible/roles/formplayer/tasks/main.yml

```diff
@@ -67,18 +67,6 @@
   tags:
     - formplayer_deploy
 
-- name: Link formplayer jar
-  copy:
-    remote_src: yes
-    src: "{{ formplayer_release_dir }}/libs/formplayer.jar"
-    dest: "{{ formplayer_build_dir }}/formplayer.jar"
-    mode: 0544
-    owner: "{{ cchq_user }}"
-    group: "{{ cchq_user }}"
-  tags:
-    - formplayer_deploy
-
-
 - name: Copy Spring application.properties
```

Comment:
Looks like Spring and Sentry properties as well as the formplayer jar are copied or linked directly into `formplayer_build_dir`. I think this means that if any of these deploy steps fails, then the build dir will potentially be in an inconsistent state (wrong properties for the linked jar, for example) and a subsequent restart of formplayer may have a bad outcome. This is the issue I was getting at in https://github.com/dimagi/commcare-cloud/pull/3253#discussion_r323863437

I was expecting the solution would be to copy all of these files into a new _release dir_, and then link that _release dir_ to the _build dir_ (maybe better named _current dir_?) as the final step. Does that sound like a good solution, or is there something I'm overlooking?

Reply:
Thank you for persistence, because you're right, I totally missed the point you were trying to make earlier, specifically the part about making sure we update the config files in a separate release as well.

I think the overall risk given our current process is low; or rather I don't think this introduces risks that aren't already in our current process. Currently, `update-config` will update the sentry and spring property files for formplayer (along with localsettings.py for commcare-hq code), without restarting formplayer. (I'm not sure if the live formplayer process then picks up these changes dynamically, but my guess is that it doesn't.) Then separately the formplayer.jar during the formplayer deploy and then restarted. (Also formplayer is currently getting restarted without the jar getting updated on every deploy as a workaround for an issue.) All that said, I do think that you're right that we can do better than that, and that with all the very-related changes in this PR, this PR might be the right time to get this right.

Something I do want to do is keep nice diffs during the update process. Probably the easiest way to achieve that would be to start by copying the files from the `current` directory, and then update them with the templated code. That seems easy enough and shouldn't add more than a couple seconds to the `formplayer deploy`.

I'll investigate this and unless it gets messy for some reason I haven't thought of, I'll aim to include the changes I come up with in this PR.

Thanks again for the suggestion and persistence!